### PR TITLE
Support encrypted synchronized Realms

### PR DESCRIPTION
**Pull request by jpsim**
Still missing changelog entries for this and
other changes that happen to have been pulled in due to the
core, sync and object store updates

### CODE REVIEW

File: Realm/RLMQueryUtil.mm [hunk 1]
```diff
@@ -126,6 +126,7 @@ size_t find_first(size_t start, size_t end) const override
         return realm::not_found;
     }
     void set_base_table(const Table*) override {}
+    void verify_column() const override {}
```
Comment (austinzheng) [inline Realm/RLMQueryUtil.mm]: I ran into this issue as well. We should make sure that an empty override is the right thing to do in this case.

File: Realm/RLMRealm.mm [hunk 2]
```diff
@@ -289,6 +290,12 @@ + (instancetype)realmWithConfiguration:(RLMRealmConfiguration *)configuration er
     configuration = [configuration copy];
     Realm::Config& config = configuration.config;
 
+    if (configuration.encryptionKey && config.sync_config) {
+        std::array<char, 64> sync_encryption_key;
+        std::copy_n(config.encryption_key.begin(), 64, sync_encryption_key.begin());
+        config.sync_config.get()->realm_encryption_key = std::move(sync_encryption_key);
```
Comment (bdash) [inline Realm/RLMRealm.mm]: `std::move` isn't useful here as the data is stored directly within `std::array` and inherently has to be copied. It'd probably be simplest to just populate the `realm_encryption_key` member directly rather than going via a temporary.

File: Realm/RLMQueryUtil.mm (hunk 1 above)
Comment (bdash) [inline Realm/RLMQueryUtil.mm]: It is. `TrueExpression` / `FalseExpression` don't have columns so there's nothing to verify.

File: Realm/RLMRealm.mm (hunk 2 above)
Comment (austinzheng) [inline Realm/RLMRealm.mm]: My bad. I suggested this approach.

**Comment by austinzheng** [comment]
Hmm...tests are failing with `fatal error: 'sync_config.hpp' file not found: #include "sync_config.hpp"`. Can you check the target membership settings on the object store files in Xcode?

File: Realm/RLMRealm.mm [hunk 3]
```diff
@@ -41,6 +41,7 @@
 #include "object_store.hpp"
 #include "schema.hpp"
 #include "shared_realm.hpp"
+#include "sync_config.hpp"
```
Comment (tgoyne) [inline Realm/RLMRealm.mm]: sync/sync_config.hpp

File: Realm/RLMRealm.mm (hunk 2 above)
Comment (tgoyne) [inline Realm/RLMRealm.mm]: Setting the encryption key on the SyncConfig object should be done in RLMRealmConfiguration like all of the other fields.

File: Realm/ObjectServerTests/RLMObjectServerTests.m [hunk 4]
```diff
@@ -318,6 +319,67 @@ - (void)testDeleteObjects {
...
+        NSString *path;
+        @autoreleasepool {
+            RLMRealm *realm = [self openRealmForURL:url user:user encryptionKey:RLMGenerateKey()
+                                         stopPolicy:RLMSyncStopPolicyImmediately immediatelyBlock:nil];
+            path = realm.configuration.pathOnDisk;
+            CHECK_COUNT(0, SyncObject, realm);
+            RLMRunChildAndWait();
+            [self waitForDownloadsForUser:user realms:@[realm] realmURLs:@[url] expectedCounts:@[@3]];
+        }
+        RLMRealmConfiguration *c = [RLMRealmConfiguration defaultConfiguration];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: This test should also test re-opening it as a synced Realm with the wrong encryption key.

Comment (jpsim) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: I'd like to do that, but the sync client is getting in the way by throwing an exception in its own thread, even with the sync policy mode set to "immediately". Do you have any ideas on how to catch that? Or make the object store / cocoa exception happen before that point?

File: Realm/RLMRealm.mm (hunk 2 above)
Comment (jpsim) [inline Realm/RLMRealm.mm]: That complicates things unnecessarily. What if the encryption key is set before the sync configuration is set?

Comment (tgoyne) [inline Realm/RLMRealm.mm]: Having a weird special case in some unrelated code is much worse than an extra few lines of code.

Comment (jpsim) [inline Realm/RLMRealm.mm]: Done.

File: Realm/RLMRealmConfiguration+Sync.mm [hunk 5]
```diff
@@ -51,6 +51,10 @@ - (void)setSyncConfiguration:(RLMSyncConfiguration *)syncConfiguration {
     self.config.in_memory = false;
     self.config.sync_config = std::make_shared<realm::SyncConfig>([syncConfiguration rawConfiguration]);
     self.config.schema_mode = realm::SchemaMode::Additive;
+    if (!self.config.encryption_key.empty()) {
+        self.config.sync_config.get()->realm_encryption_key = std::array<char, 64>();
```
Comment (bdash) [inline Realm/RLMRealmConfiguration+Sync.mm]: `.get()->` on a `unique_ptr` / `shared_ptr` is equivalent to just `->`.

I'd probably add `auto& sync_encryption_key = self.config.sync_config->realm_encryption_key;` here so that this line and the following can refer to it as `sync_encryption_key`.

File: Realm/RLMRealmConfiguration.mm [hunk 6]
```diff
@@ -25,6 +25,9 @@
 
 #import "schema.hpp"
 #import "shared_realm.hpp"
+#import "sync/sync_config.hpp"
+
+#include <realm/util/optional.hpp>
```
Comment (bdash) [inline Realm/RLMRealmConfiguration.mm]: This doesn't appear to be necessary. `SyncConfig` has a `util::Optional` as a member so is required to have included this header.

File: Realm/RLMRealmConfiguration.mm [hunk 7]
```diff
@@ -176,9 +177,16 @@ - (void)setEncryptionKey:(NSData * __nullable)encryptionKey {
...
+        if (_config.sync_config) {
+            auto& sync_encryption_key = self.config.sync_config->realm_encryption_key;
+            sync_encryption_key = std::array<char, 64>();
+            std::copy_n(_config.encryption_key.begin(), 64, sync_encryption_key->begin());
+        }
     }
     else {
         _config.encryption_key.clear();
+        if (_config.sync_config)
+            _config.sync_config.get()->realm_encryption_key = realm::util::none;
```
Comment (bdash) [inline Realm/RLMRealmConfiguration.mm]: You overlooked this `.get()->`

File: dependencies.list [hunk 8]
```diff
@@ -1,4 +1,4 @@
 VERSION=2.4.4
-REALM_CORE_VERSION=2.3.3
-REALM_SYNC_VERSION=1.3.1
-REALM_OBJECT_SERVER_VERSION=1.2.1
+REALM_CORE_VERSION=2.4.0
+REALM_SYNC_VERSION=1.4.2
```
Comment (bdash) [inline dependencies.list]: Is there a reason we're going to 1.4.2 here rather than 1.5.0?

Comment (jpsim) [inline dependencies.list]: Not that I can think of, no. I'll update.

**Comment by jpsim** [comment]
I've added some changelog entries. I don't think realm/realm-core#2415 was ever triggerable by our end users, so I didn't include that in the changelog entries, but I'd appreciate if someone could confirm.

**Comment by tgoyne** [comment]
realm/realm-core#2415 is only relevant with a true dynamic schema, which we don't support.

File: Realm/ObjectServerTests/RLMObjectServerTests.m (hunk 4 above)
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: I suppose it's fine for now if we need to get this out, but crashing in an awkward place when given an incorrect encryption key sounds like a problem that needs to be addressed.

Comment (jpsim) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: Is this a deal breaker for you to get this in?

Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: I suppose it's fine for now if we need to get this out.

Fixing it probably requires changes on the sync side to report the error in a better way.

**Comment by jpsim** [comment]
Addressed all feedback so far, have changelog entries, updated to core 2.5.1 and sync 1.5.0. Object store is at realm/realm-object-store#420, which I'll update to point to `master` once that PR passes CI and is merged.

File: CHANGELOG.md [hunk 9]
```diff
@@ -11,6 +17,8 @@ x.x.x Release notes (yyyy-MM-dd)
 * Add support for the `[d]` modifier on string comparison operators to perform
   diacritic-insensitive comparisons.
 * Explicitly mark `[[RLMRealm alloc] init]` as unavailable.
+* Don't unnecessarily advance the Realm transaction when resolving a
```
Comment (bdash) [inline CHANGELOG.md]: This reads more like a bug fix than an enhancement?

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,6 +1,12 @@
 x.x.x Release notes (yyyy-MM-dd)
 =============================================================
 
+This release requires all processes accessing a Realm file concurrently to be on
+this version or later. This includes iOS extensions and the Realm Browser.
```
Comment (tgoyne) [inline CHANGELOG.md]: I believe older versions will fail to open files written by this version even if it's non-concurrent.

File: CHANGELOG.md (hunk 9 above)
Comment (tgoyne) [inline CHANGELOG.md]: I'm not sure what change this is referring to? 7cf4c6f is the only TSR-related change that I see, and that's something significantly different that doesn't apply to cocoa.

Comment (jpsim) [inline CHANGELOG.md]: Yes, that's the commit that prompted me to write this entry, but you're right that it doesn't apply. Removed.
